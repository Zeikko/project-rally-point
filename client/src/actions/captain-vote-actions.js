import * as actions from '../../../common/actions.json'

export const voteCaptainAction = (gameId, playerId, userId) => ({
  type: actions.CAPTAIN_VOTE_REQUEST,
  socket: true,
  playerId,
  gameId,
  userId,
})
