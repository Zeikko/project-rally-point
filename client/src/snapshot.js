import renderer from 'react-test-renderer'
import 'jest-styled-components'

export function matchComponentToSnapshot(component) {
  const json = renderer.create(component).toJSON()
  expect(json).toMatchSnapshot()
}