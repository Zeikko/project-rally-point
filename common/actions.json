{
  "GET_USER_REQUEST": "GET_USER_REQUEST",
  "GET_USER_SUCCESS": "GET_USER_SUCCESS",
  "GET_USER_ERROR": "GET_USER_ERROR",

  "LOGOUT_REQUEST": "LOGOUT_REQUEST",
  "LOGOUT_SUCCESS":"LOGOUT_SUCCESS",
  "LOGOUT_ERROR": "LOGOUT_ERROR",

  "JOIN_GAME_REQUEST": "JOIN_GAME_REQUEST",
  "JOIN_GAME_SUCCESS": "JOIN_GAME_SUCCESS",
  "JOIN_GAME_ERROR": "JOIN_GAME_ERROR",

  "LEAVE_GAME_REQUEST": "LEAVE_GAME_REQUEST",
  "LEAVE_GAME_SUCCESS": "LEAVE_GAME_SUCCESS",
  "LEAVE_GAME_ERROR": "LEAVE_GAME_ERROR",

  "GET_GAME_REQUEST": "GET_GAME_REQUEST",
  "GET_GAME_SUCCESS": "GET_GAME_SUCCESS",
  "GET_GAME_ERROR": "GET_GAME_ERROR",

  "GET_PLAYERS_REQUEST": "GET_PLAYERS_REQUEST",
  "GET_PLAYERS_SUCCESS": "GET_PLAYERS_SUCCESS",
  "GET_PLAYERS_ERROR": "GET_PLAYERS_ERROR",

  "CAPTAIN_VOTE_REQUEST": "CAPTAIN_VOTE_REQUEST",
  "CAPTAIN_VOTE_SUCCESS": "CAPTAIN_VOTE_SUCCESS",
  "CAPTAIN_VOTE_ERROR": "CAPTAIN_VOTE_ERROR",

  "GET_CAPTAIN_VOTES_REQUEST": "GET_CAPTAIN_VOTES_REQUEST",
  "GET_CAPTAIN_VOTES_SUCCESS": "GET_CAPTAIN_VOTES_SUCCESS",
  "GET_CAPTAIN_VOTES_ERROR": "GET_CAPTAIN_VOTES_ERROR",

  "CREATE_GAME_REQUEST": "CREATE_GAME_REQUEST",
  "CREATE_GAME_SUCCESS": "CREATE_GAME_SUCCESS",
  "CREATE_GAME_ERROR": "CREATE_GAME_ERROR",

  "PICK_PLAYER_REQUEST": "PICK_PLAYER_REQUEST",
  "PICK_PLAYER_SUCCESS": "PICK_PLAYER_SUCCESS",
  "PICK_PLAYER_ERROR": "PICK_PLAYER_ERROR",

  "SIMULATE_USER": "SIMULATE_USER"

}
