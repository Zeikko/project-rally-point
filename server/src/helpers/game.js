import _ from 'lodash'
import db from '../db'
import gameStatuses from '../../../common/game-statuses.json'
import playerRoles from '../../../common/player-roles.json'

export async function shouldStartCaptainVote(gameId) {
  const game = await db('game')
    .first('*')
    .where({ id: gameId })
  const numberOfPlayers = await db('player')
    .count('id')
    .where({ gameId })
  return parseInt(numberOfPlayers[0].count, 10) === game.maxPlayers
}

export async function shouldStartSquadLeaderPick(gameId) {
  const game = await db('game')
    .first('*')
    .where({ id: gameId })
  const numberOfCaptainVotes = await db('captainVote')
    .count('id')
    .where({ gameId })
  return parseInt(numberOfCaptainVotes[0].count, 10) === game.maxPlayers
}

export async function shouldStartSquadMemberPick(gameId) {
  const game = await db('game')
    .first('*')
    .where({ id: gameId })
  const numberOfSquadLeaders = await db('player')
    .select('id')
    .where({
      gameId,
      role: playerRoles.SQUAD_LEADER,
      team: game.teamWithTurnToPick,
    })
  return (numberOfSquadLeaders.length + 1) === getOptimalNumberOfSquads(game)
}

export async function shouldStartPlayingGame(gameId) {
  const game = await db('game')
    .first('*')
    .where({ id: gameId })
  const numberOfPlayers = await db('player')
    .count('id')
    .where({ gameId })
    .whereNotNull('team')
  return parseInt(numberOfPlayers[0].count, 10) === game.maxPlayers
}

export async function startCaptainVote(gameId) {
  const games = await db('game')
    .update({ status: gameStatuses.VOTE_CAPTAINS })
    .where({ id: gameId })
    .returning('*')
  return _.first(games)
}

export async function startSquadLeaderPick(gameId) {
  const games = await db('game')
    .update({
      status: gameStatuses.SQUAD_LEADER_PICK,
      teamWithTurnToPick: 1,
    })
    .where({ id: gameId })
    .returning('*')
  return _.first(games)
}

export async function startSquadMemberPick(gameId) {
  const games = await db('game')
    .update({ status: gameStatuses.SQUAD_MEMBER_PICK })
    .where({ id: gameId })
    .returning('*')
  return _.first(games)
}

export async function startPlayingGame(gameId) {
  const games = await db('game')
    .update({ status: gameStatuses.PLAYING })
    .where({ id: gameId })
    .returning('*')
  return _.first(games)
}

export function getGameStatus(gameId) {
  return db('game')
    .first('status')
    .where({ id: gameId })
}

export async function passSquadLeaderPickTurn(game) {
  const games = await db('game')
    .update({
      teamWithTurnToPick: (game.teamWithTurnToPick % 2) + 1,
    })
    .where({ id: game.id })
    .returning('*')
  return _.first(games)
}

export async function passPlayerPickTurn(game, playersCount) {
  const nextTeamWithTurnToPick = await calculatePickTurn(game, playersCount)
  const games = await db('game')
    .update({
      teamWithTurnToPick: nextTeamWithTurnToPick,
    })
    .where({ id: game.id })
    .returning('*')
  return _.first(games)
}

async function calculatePickTurn(game, playersCount) {
  const numberOfSquads = getOptimalNumberOfSquads(game)
  if (playersCount % numberOfSquads === 0) {
    return (game.teamWithTurnToPick % 2) + 1
  }
  return game.teamWithTurnToPick
}

export function getOptimalNumberOfSquads(game) {
  return Math.ceil(game.maxPlayers / 2 / 9)
}
