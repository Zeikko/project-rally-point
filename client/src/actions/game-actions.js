import * as actions from '../../../common/actions.json'

export const joinGameAction = (id) => ({
  type: actions.JOIN_GAME_REQUEST,
  socket: true,
  id
})

export const leaveGameAction = (id) => ({
  type: actions.LEAVE_GAME_REQUEST,
  socket: true,
  id
})
