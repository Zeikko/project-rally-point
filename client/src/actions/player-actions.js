import * as actions from '../../../common/actions.json'

export const joinGameAction = (gameId, userId) => ({
  type: actions.JOIN_GAME_REQUEST,
  socket: true,
  gameId,
  userId,
})

export const leaveGameAction = (gameId, userId) => ({
  type: actions.LEAVE_GAME_REQUEST,
  socket: true,
  gameId,
  userId,
})
