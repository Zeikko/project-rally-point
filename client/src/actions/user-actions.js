import actions from '../../../common/actions.json'

export const getUserAction = () => ({
  type: actions.GET_USER_REQUEST,
})

export const logoutAction = () => ({
  type: actions.LOGOUT_REQUEST,
})
