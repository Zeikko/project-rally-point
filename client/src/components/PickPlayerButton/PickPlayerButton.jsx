import _ from 'lodash'
import PropTypes from 'prop-types'
import React, { Component } from 'react'
import { pickSquadLeaderAction, pickSquadMemberAction } from '../../actions/player-actions'
import Button from '../Button/Button'
import gameStatuses from '../../../../common/game-statuses.json'
import playerRoles from '../../../../common/player-roles.json'
import * as propTypes from '../../constants/prop-types'

class PickPlayerButton extends Component {
  constructor() {
    super()
    this.handleClick = this.handleClick.bind(this)
  }

  handleClick() {
    const {
      dispatch, game, player, user,
    } = this.props
    if (game.status === gameStatuses.SQUAD_LEADER_PICK) {
      dispatch(pickSquadLeaderAction(game.id, player.id, game.teamWithTurnToPick, user.id))
    } else {
      dispatch(pickSquadMemberAction(game.id, player.id, game.teamWithTurnToPick, user.id))
    }
  }

  render() {
    const { game, players, user } = this.props
    const userPlayer = _.find(players, { id: user.id })
    if (userPlayer) {
      const userPlayerRole = _.get(userPlayer, 'role')
      const userPlayerTeam = _.get(userPlayer, 'team')
      const isUserPlayersTurn = game.teamWithTurnToPick === userPlayerTeam
      const isCaptain = userPlayerRole === playerRoles.CAPTAIN
      const isSquadLeader = userPlayerRole === playerRoles.SQUAD_LEADER
      const playersInTeam = _.filter(players, { team: userPlayer.team })
      const playerCountInSquad = _.filter(playersInTeam, { squad: userPlayer.squad }).length
      const smallestSquadPlayerCount = _.chain(playersInTeam)
        .groupBy(player => player.squad)
        .toArray()
        .minBy(squad => squad.length)
        .get('length', 0)
        .value()
      const canPickSquadLeader = game.status === gameStatuses.SQUAD_LEADER_PICK
        && isCaptain
        && isUserPlayersTurn
      const canPickSquadMember = game.status === gameStatuses.SQUAD_MEMBER_PICK
        && (isCaptain || isSquadLeader)
        && isUserPlayersTurn
        && playerCountInSquad <= smallestSquadPlayerCount
      if (canPickSquadLeader || canPickSquadMember) {
        return (
          <Button onClick={() => { this.handleClick() }}>
            Pick Player
          </Button>
        )
      }
    }
    return null
  }
}

PickPlayerButton.propTypes = {
  dispatch: PropTypes.func.isRequired,
  player: propTypes.player.isRequired,
  game: propTypes.game.isRequired,
  players: PropTypes.arrayOf(propTypes.player).isRequired,
  user: propTypes.user.isRequired,
}

export default PickPlayerButton
