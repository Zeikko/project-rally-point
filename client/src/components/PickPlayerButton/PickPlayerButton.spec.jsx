import React from 'react'
import { mount } from 'enzyme'
import { matchComponentToSnapshot } from '../../test/snapshot'
import PickPlayerButton from './PickPlayerButton'
import normalUser from '../../../../fixtures/normal-user.json'
import game from '../../../../fixtures/game.json'
import player from '../../../../fixtures/player.json'
import gameStatuses from '../../../../common/game-statuses.json'
import playerRoles from '../../../../common/player-roles.json'
import actions from '../../../../common/actions.json'

describe('PickPlayerButton', () => {
  it('does not render anything when it is not time to pick player', () => {
    const players = [player]
    matchComponentToSnapshot(<PickPlayerButton
      dispatch={() => {}}
      player={player}
      game={game}
      players={players}
      user={normalUser}
    />)
  })

  it('renders button to pick a captain', () => {
    const captainPlayer = { ...player, team: 1, role: playerRoles.CAPTAIN }
    const players = [captainPlayer]
    matchComponentToSnapshot(<PickPlayerButton
      dispatch={() => {}}
      player={captainPlayer}
      game={{ ...game, teamWithTurnToPick: 1, status: gameStatuses.SQUAD_LEADER_PICK }}
      players={players}
      user={{ ...normalUser, id: 1 }}
    />)
  })

  it('picks a captain', () => {
    const dispatch = jest.fn()
    const captainPlayer = { ...player, team: 1, role: playerRoles.CAPTAIN }
    const players = [captainPlayer]
    const wrapper = mount(<PickPlayerButton
      dispatch={dispatch}
      player={captainPlayer}
      game={{ ...game, teamWithTurnToPick: 1, status: gameStatuses.SQUAD_LEADER_PICK }}
      players={players}
      user={{ ...normalUser, id: 1 }}
    />)
    wrapper.find('button').simulate('click')
    expect(dispatch.mock.calls[0]).toEqual([
      {
        type: actions.PICK_SQUAD_LEADER_REQUEST,
        socket: true,
        gameId: 1,
        team: 1,
        userId: 1,
        playerId: 1,
      },
    ])
  })
})
