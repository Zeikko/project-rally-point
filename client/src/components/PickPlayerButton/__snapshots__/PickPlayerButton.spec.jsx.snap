// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PickPlayerButton does not render anything when it is not time to pick player 1`] = `null`;

exports[`PickPlayerButton does not render when user is not in the game 1`] = `null`;

exports[`PickPlayerButton renders button to pick a captain 1`] = `
.c0 {
  font-size: 18px;
  border: solid black 1px;
  background-color: #AAA;
  cursor: pointer;
}

<button
  className="c0"
  onClick={[Function]}
>
  Pick Player
</button>
`;

exports[`PickPlayerButton renders button to pick a squad member 1`] = `
.c0 {
  font-size: 18px;
  border: solid black 1px;
  background-color: #AAA;
  cursor: pointer;
}

<button
  className="c0"
  onClick={[Function]}
>
  Pick Player
</button>
`;
